File: domains/games/apis/one_d4/src/test/java/com/muchq/games/one_d4/worker/OpeningsTest.java
package com.muchq.games.one_d4.worker;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

public class OpeningsTest {

  @Test
  public void nameFromEcoUrl_extractsSlugAsSpacedName() {
    assertThat(
            Openings.nameFromEcoUrl(
                "https://www.chess.com/openings/Caro-Kann-Defense-Two-Knights-Attack-3...dxe4"))
        .isEqualTo("Caro Kann Defense Two Knights Attack 3...dxe4");
    assertThat(
            Openings.nameFromEcoUrl(
                "https://www.chess.com/openings/English-Opening-Agincourt-Defense-2.Nf3-d5-3.g3"))
        .isEqualTo("English Opening Agincourt Defense 2.Nf3 d5 3.g3");
  }

  /**
   * The move continuation belongs in the name — it is what makes opening_name the fine-grained
   * value that opening_family is the coarse counterpart of. The #1344 strip is scoped to
   * familyFromName; hoisting it up here would silently coarsen opening_name too.
   */
  @Test
  public void nameFromEcoUrl_keepsTheMoveContinuation() {
    assertThat(
            Openings.nameFromEcoUrl(
                "https://www.chess.com/openings/Owens-Defense...3.Nc3-e6-4.Nf3-Bb4"))
        .isEqualTo("Owens Defense...3.Nc3 e6 4.Nf3 Bb4");
  }

  @Test
  public void nameFromEcoUrl_toleratesTrailingSlash() {
    assertThat(Openings.nameFromEcoUrl("https://www.chess.com/openings/Vienna-Game/"))
        .isEqualTo("Vienna Game");
  }

  @Test
  public void nameFromEcoUrl_bareEcoCodeYieldsNull() {
    assertThat(Openings.nameFromEcoUrl("B10")).isNull();
    assertThat(Openings.nameFromEcoUrl("https://www.chess.com/openings/B10")).isNull();
  }

  @Test
  public void nameFromEcoUrl_nullOrBlankYieldsNull() {
    assertThat(Openings.nameFromEcoUrl(null)).isNull();
    assertThat(Openings.nameFromEcoUrl("  ")).isNull();
    assertThat(Openings.nameFromEcoUrl("https://www.chess.com/openings/")).isNull();
  }

  @Test
  public void familyFromName_stopsAtFirstStructuralWord() {
    assertThat(Openings.familyFromName("English Opening Agincourt Defense 2.Nf3 d5 3.g3"))
        .isEqualTo("English Opening");
    assertThat(Openings.familyFromName("Caro Kann Defense Two Knights Attack 3...dxe4"))
        .isEqualTo("Caro Kann Defense");
    assertThat(Openings.familyFromName("Kings Indian Attack")).isEqualTo("Kings Indian Attack");
    assertThat(Openings.familyFromName("Queens Gambit Declined Modern Variation"))
        .isEqualTo("Queens Gambit");
    assertThat(Openings.familyFromName("London System")).isEqualTo("London System");
  }

  /**
   * #1344: chess.com glues the move continuation onto the *first* structural word, hiding it from
   * the terminator scan. Both halves of that failure are here — the raw moves leaking into the key
   * ("Owens Defense...3.Nc3"), and the family silently losing its structural word ("Giuoco Piano"
   * for "Giuoco Piano Game"), which is the worse one: it splits a single family across two
   * opening_family group keys, so per-family counts and win rates are wrong with nothing in the
   * output to show it. The last case is the corpus's castling shape, where "O-O" despaces into "O
   * O" and the two-word fallback cut mid-continuation ("Pirc Defense...6.O").
   */
  @Test
  public void familyFromName_stripsAMoveContinuationOnTheFirstStructuralWord() {
    assertThat(Openings.familyFromName("Owens Defense...3.Nc3 e6 4.Nf3 Bb4"))
        .isEqualTo("Owens Defense");
    assertThat(Openings.familyFromName("Giuoco Piano Game...5.d3 d6 6.c3 O O 7.Re1"))
        .isEqualTo("Giuoco Piano Game");
    assertThat(Openings.familyFromName("Kings Indian Attack...3.Bg2 e5 4.d3 Nf6 5.O O"))
        .isEqualTo("Kings Indian Attack");
    assertThat(Openings.familyFromName("Pirc Defense...6.O O Bg7 7.h3 O O"))
        .isEqualTo("Pirc Defense");
  }

  /**
   * The shape that already worked before #1344 — the continuation rides a later word, so the scan
   * cut before ever reaching it. It is why OpeningsTest passed while real months were mis-bucketed,
   * so it stays pinned alongside the shape that didn't.
   */
  @Test
  public void familyFromName_stripsAMoveContinuationOnALaterWord() {
    assertThat(Openings.familyFromName("Sicilian Defense Chekhover Variation...7.Nc3 Nf6 8.Bg5 e6"))
        .isEqualTo("Sicilian Defense");
    assertThat(Openings.familyFromName("Caro Kann Defense Two Knights Attack 3...dxe4"))
        .isEqualTo("Caro Kann Defense");
  }

  /**
   * Cut at the first continuation, not the last. chess.com writes a black-to-move continuation as
   * "4...Bb4", so a slug can carry a second "..." inside the moves, and cutting at the last one
   * leaves the earlier moves in the key. Nothing in the frozen corpus has two — this is the
   * mutation the corpus suite could not kill, pinned by hand rather than left to intent.
   */
  @Test
  public void familyFromName_cutsAtTheFirstOfSeveralMoveContinuations() {
    assertThat(Openings.familyFromName("Owens Defense...3.Nc3 e6 4...Bb4"))
        .isEqualTo("Owens Defense");
  }

  @Test
  public void familyFromName_fallsBackToFirstTwoWords() {
    assertThat(Openings.familyFromName("Something Unusual Line Here"))
        .isEqualTo("Something Unusual");
    assertThat(Openings.familyFromName("Singleword")).isEqualTo("Singleword");
  }

  /** The fallback counts words of the stripped name, not of the continuation. */
  @Test
  public void familyFromName_fallsBackToFirstTwoWordsOfTheStrippedName() {
    assertThat(Openings.familyFromName("Grob...2.Bg2 e5")).isEqualTo("Grob");
    assertThat(Openings.familyFromName("Something Unusual...2.Nf3 d5 3.g3"))
        .isEqualTo("Something Unusual");
  }

  @Test
  public void familyFromName_nullOrBlankYieldsNull() {
    assertThat(Openings.familyFromName(null)).isNull();
    assertThat(Openings.familyFromName(" ")).isNull();
  }

  /** Nothing survives the strip, so there is no family — not an empty-string group key. */
  @Test
  public void familyFromName_moveContinuationOnlyYieldsNull() {
    assertThat(Openings.familyFromName("...3.Nc3 e6")).isNull();
    assertThat(Openings.familyFromName("  ...  ")).isNull();
  }
}

File: domains/games/apis/one_d4_worker/openings_test.cc
#include "domains/games/apis/one_d4_worker/openings.h"

#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <fstream>
#include <regex>
#include <set>
#include <sstream>
#include <string>
#include <vector>

namespace one_d4_worker {
namespace {

using ::testing::Contains;
using ::testing::Not;

TEST(Openings, ReadsTheSlugAsASpacedName) {
  EXPECT_EQ(OpeningNameFromEcoUrl("https://www.chess.com/openings/Vienna-Game-Max-Lange-Defense"),
            "Vienna Game Max Lange Defense");
  EXPECT_EQ(OpeningNameFromEcoUrl("https://www.chess.com/openings/Vienna-Game/"), "Vienna Game");
}

TEST(Openings, KeepsTheMoveContinuationInTheName) {
  // The continuation is dropped for the family only. Hoisting that up here
  // would silently coarsen opening_name too.
  EXPECT_EQ(OpeningNameFromEcoUrl(
                "https://www.chess.com/openings/Caro-Kann-Defense-Two-Knights-Attack-3...dxe4"),
            "Caro Kann Defense Two Knights Attack 3...dxe4");
}

TEST(Openings, HasNoNameForACodeOrABlank) {
  EXPECT_EQ(OpeningNameFromEcoUrl("B10"), "");
  EXPECT_EQ(OpeningNameFromEcoUrl("https://www.chess.com/openings/B10"), "");
  EXPECT_EQ(OpeningNameFromEcoUrl(""), "");
  EXPECT_EQ(OpeningNameFromEcoUrl("  "), "");
  EXPECT_EQ(OpeningNameFromEcoUrl("https://www.chess.com/openings/"), "");
}

TEST(Openings, FamilyStopsAtTheFirstStructuralWord) {
  EXPECT_EQ(OpeningFamilyFromName("English Opening Agincourt Defense 2.Nf3 d5 3.g3"),
            "English Opening");
  EXPECT_EQ(OpeningFamilyFromName("Caro Kann Defense Two Knights Attack 3...dxe4"),
            "Caro Kann Defense");
  EXPECT_EQ(OpeningFamilyFromName("Kings Indian Attack"), "Kings Indian Attack");
  EXPECT_EQ(OpeningFamilyFromName("Queens Gambit Declined Modern Variation"), "Queens Gambit");
  EXPECT_EQ(OpeningFamilyFromName("London System"), "London System");
}

TEST(Openings, FamilyDropsAContinuationGluedToTheStructuralWord) {
  // "Owens Defense...3.Nc3" splits as one word, so a scan that does not cut
  // the continuation first never sees "Defense" and truncates to two words.
  EXPECT_EQ(OpeningFamilyFromName("Owens Defense...3.Nc3 e6 4.Nf3 Bb4"), "Owens Defense");
  EXPECT_EQ(OpeningFamilyFromName("Giuoco Piano Game...5.d3 d6 6.c3 O O 7.Re1"),
            "Giuoco Piano Game");
  EXPECT_EQ(OpeningFamilyFromName("Kings Indian Attack...3.Bg2 e5"), "Kings Indian Attack");
  EXPECT_EQ(OpeningFamilyFromName("Sicilian Defense Chekhover Variation...7.Nc3 Nf6"),
            "Sicilian Defense");
  EXPECT_EQ(OpeningFamilyFromName("Owens Defense...3.Nc3 e6 4...Bb4"), "Owens Defense");
}

TEST(Openings, FamilyFallsBackToTheFirstTwoWords) {
  EXPECT_EQ(OpeningFamilyFromName("Something Unusual Line Here"), "Something Unusual");
  EXPECT_EQ(OpeningFamilyFromName("Singleword"), "Singleword");
  EXPECT_EQ(OpeningFamilyFromName("Grob...2.Bg2 e5"), "Grob");
  EXPECT_EQ(OpeningFamilyFromName("Something Unusual...2.Nf3 d5 3.g3"), "Something Unusual");
}

TEST(Openings, HasNoFamilyForABlankOrAContinuationAlone) {
  EXPECT_EQ(OpeningFamilyFromName(""), "");
  EXPECT_EQ(OpeningFamilyFromName(" "), "");
  EXPECT_EQ(OpeningFamilyFromName("...3.Nc3 e6"), "");
  EXPECT_EQ(OpeningFamilyFromName("  ...  "), "");
}

// --- the same corpus the Java derivation is held to ------------------------

// OpeningsCorpusTest's numbers, over the same frozen file. They are the
// positive control: a derivation that stops collapsing continuations moves
// the family count (52 before #1344, 40 after) even when every individual
// key still looks clean.
constexpr int kEcoUrls = 500;
constexpr std::size_t kDistinctSlugs = 307;
constexpr std::size_t kDistinctFamilies = 40;

std::vector<std::string> CorpusEcoUrls() {
  std::ifstream file("domains/games/apis/one_d4/src/test/resources/hikaru_corpus.pgn");
  EXPECT_TRUE(file.good()) << "hikaru_corpus.pgn is not where this test looks";
  std::ostringstream contents;
  contents << file.rdbuf();
  const std::string text = contents.str();

  std::vector<std::string> urls;
  const std::regex tag(R"re(\[ECOUrl "([^"]+)"\])re");
  for (std::sregex_iterator it(text.begin(), text.end(), tag), end; it != end; ++it) {
    urls.push_back((*it)[1]);
  }
  return urls;
}

std::set<std::string> CorpusFamilies() {
  const std::vector<std::string> urls = CorpusEcoUrls();
  EXPECT_EQ(urls.size(), static_cast<std::size_t>(kEcoUrls));
  EXPECT_EQ(std::set<std::string>(urls.begin(), urls.end()).size(), kDistinctSlugs);

  std::set<std::string> families;
  for (const std::string& url : urls) {
    const std::string name = OpeningNameFromEcoUrl(url);
    EXPECT_FALSE(name.empty()) << "no name for " << url;
    const std::string family = OpeningFamilyFromName(name);
    EXPECT_FALSE(family.empty()) << "no family for " << url;
    families.insert(family);
  }
  return families;
}

TEST(Openings, NoCorpusFamilyCarriesAMoveContinuation) {
  // Digits and '.' are the tell: no opening name carries either.
  for (const std::string& family : CorpusFamilies()) {
    EXPECT_EQ(family.find('.'), std::string::npos) << family;
    EXPECT_EQ(family.find_first_of("0123456789"), std::string::npos) << family;
  }
}

TEST(Openings, DerivesTheSameFamiliesAsTheJavaWorker) {
  const std::set<std::string> families = CorpusFamilies();
  EXPECT_EQ(families.size(), kDistinctFamilies);
  EXPECT_THAT(families, Contains("Giuoco Piano Game"));
  EXPECT_THAT(families, Contains("Kings Indian Attack"));
  EXPECT_THAT(families, Contains("Old Benoni Defense"));
  EXPECT_THAT(families, Contains("Owens Defense"));
  EXPECT_THAT(families, Not(Contains("Giuoco Piano")));
  EXPECT_THAT(families, Not(Contains("Kings Indian")));
  EXPECT_THAT(families, Not(Contains("Old Benoni")));
}

}  // namespace
}  // namespace one_d4_worker
